Implement ECH-10.

chris-wood opened the pull request:
This change adds support for ECH-10. The primary changes are moving to
HPKE-08 and changing the ECHConfig identifier from a client-computed
value to a server-chosen value. ECHProviders MUST use rejection sampling
in choosing the configuration identifier so as to not introduce conflicts.

Commit:
crypto/tls: Move draft-ietf-tls-esni-09 to 10

This change adds support for ECH-10 and removes support for ECH-09. The
primary changes are moving to HPKE-08 and changing the ECHConfig
identifier from a client-computed value to a server-chosen value.
ECHProviders MUST use rejection sampling in choosing the configuration
identifier so as to not introduce conflicts.

Lekensteyn commented on src/circl/hpke/vectors_test.go:
```diff
@@ -15,8 +15,8 @@ import (
-	// https://github.com/cfrg/draft-irtf-cfrg-hpke/blob/draft-irtf-cfrg-hpke-07/test-vectors.json
-	vectors := readFile(t, "testdata/vectors_v07.json")
+	// https://github.com/cfrg/draft-irtf-cfrg-hpke/blob/master/test-vectors.json
+	vectors := readFile(t, "testdata/vectors_v08_779d028.json")
```
Did something go wrong with checking in updated files? vectors_v07.json was updated while the new name from https://github.com/cloudflare/circl/pull/222 is not present.

chris-wood replied:
Hmm... I'm not sure I understand the question here. I just applied a patch from CIRCL's change.

Lekensteyn replied:
Sorry I should have been clearer, this vectors_test.go file is fine, however the actual test vector itself has the wrong filename:
```
$ git log -p --stat -w -n1 origin/caw/ech-10
...
 src/circl/hpke/testdata/vectors_v07.json |  2 +-
```
Try `git mv src/circl/hpke/testdata/vectors_v07.json src/circl/hpke/testdata/vectors_v08_779d028.json`

chris-wood replied:
🤦 oops. Thanks for catching that!

Lekensteyn commented:
I'll be basing the next review on the spec changelog as well as these diffs:

* https://tools.ietf.org/rfcdiff?url2=draft-ietf-tls-esni-10.txt
* https://tools.ietf.org/rfcdiff?url2=draft-irtf-cfrg-hpke-08.txt
  (Hm, draft-ietf-tls-esni-10 references the wrong -07 draft version... ok I will assume -08)

Lekensteyn commented on src/crypto/tls/conn.go:
```diff
@@ -133,6 +133,7 @@ type Conn struct {
 		innerRandom []byte         // ClientHelloInner.random
 		publicName  string         // ECHConfig.contents.public_name
+		configId    uint8          // ECHConfig.contents.key_config.config_id
 		suite       echCipherSuite // ClientECH.cipher_suite
```
Throughout the project, the type needs a rename:
```suggestion
		suite       hpkeSymmetricCipherSuite // ClientECH.cipher_suite
```

Lekensteyn commented on src/crypto/tls/ech_config.go:
```diff
@@ -87,13 +88,12 @@ func echMarshalConfigs(configs []ECHConfig) ([]byte, error) {
-		len(t)%4 != 0 {
+		!t.ReadBytes(&config.rawPublicKey, len(t)) ||
+		!contents.ReadUint16LengthPrefixed(&t) {
+		fmt.Println("here")
```
Issues:
- Excess debug print
- Missing check that `cipher_suites` is actually a multiple of 4 (the `len(t)%4 != 0` check needs to be added to the end).
- Consider adding a comment here that `HpkeKeyConfig` is being parsed here (`cipher_suites` will be parsed below though).

chris-wood replied:
Hah oops, missed this one.

chris-wood replied:
I didn't add clarifying comments since the intent seems clear as-is. Addressed the rest, though!

Lekensteyn commented on src/crypto/tls/ech_config.go:
```diff
@@ -108,9 +108,12 @@ func readConfigContents(contents *cryptobyte.String, config *ECHConfig) bool {
 		!contents.ReadUint16LengthPrefixed(&t) ||
 		!t.ReadBytes(&config.ignoredExtensions, len(t)) ||
 		!contents.Empty() {
+		fmt.Println("here2")
```
excess debug print

Lekensteyn commented on src/crypto/tls/ech_provider.go:
```diff
@@ -94,37 +90,24 @@ type EXP_ECHKeySet struct {
-			copy(id[:], b.BytesOrPanic())
-			keySet.sk[id] = key
+		if _, ok := keySet.sk[key.config.configId]; ok {
+			return nil, fmt.Errorf("tls: ech provider:: ECH config conflict")
```
Remove excess `:`, consider adding more context:
```suggestion
			return nil, fmt.Errorf("tls: ech provider: ECH config conflict for configId %d", key.config.configId)
```

Lekensteyn reviewed (changes requested):
Initial pass, it looks mostly reasonable, see inline comments for details.

Extra comments not covered elsewhere:
* Please scan for `draft-ietf-tls-esni-09` throughout the code and update references.

Lekensteyn reviewed (approved):
LGTM now, did this pass local tests?

(unfortunately we still have not configured CI yet, https://github.com/cloudflare/go/issues/27)

chris-wood commented:
> LGTM now, did this pass local tests?

Yep, though I did only run tests in the src/crypto/tls package.

claucece commented:
LGTM. When you are ready to merge, could you make the commit message of this type: `crypto/tls: What is done`, as defined in https://github.com/cloudflare/go/wiki/Contributing#all-prs

Thanks!

cjpatton committed:
circl: Update to 3977848c88c641772e447c63a0ec29c8f4085e58

This change upgrades CIRCL's implementation of draft-irtf-cfrg-hpke-07
to 08. This breaks unit tests for ECH in crypto/tls, which currently
expects HPKE-07. It also modifies the implementation of OPRF.

Manually removed Ristretto code, which includes external dependencies
that break ./all.bash.

cjpatton commented:
Rebased and split the commit into a CIRCL upgrade and a change to crypto/tls.

cjpatton reviewed (approved):
I've confirmed interop with NSS: https://github.com/xvzcf/tls-interop-runner/pull/44

Time to land.
